File: leet56/leet56.py
# https://leetcode.com/problems/merge-intervals/description/
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []
        intervals.sort()
        merged = [intervals[0]]
        for i in range(1, len(intervals)):
            last = merged[-1]
            if intervals[i][0] <= last[1]:
                last[1] = max(last[1], intervals[i][1])
            else:
                merged.append(intervals[i])
        return merged

File: leet56/leet56.cpp
// https://leetcode.com/problems/merge-intervals/description/

#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    vector<vector<int>> merge(vector<vector<int>> &intervals)
    {
        if (intervals.empty())
        {
            return {};
        }
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> merged = {intervals[0]};

        for (int i = 1; i < intervals.size(); i++)
        {
            vector<int> &last = merged.back();

            if (intervals[i][0] <= last[1])
            {
                last[1] = max(last[1], intervals[i][1]);
            }
            else
            {
                merged.push_back(intervals[i]);
            }
        }
        return merged;
    }
};